File: Cpp/396_Rotate_Function.cpp
class Solution {
public:
    int maxRotateFunction(vector<int>& A) {
        if(A.empty()) return 0;
        
        int n = A.size();
        long result = INT_MIN;
        long sum = 0;
        long rotation = 0;
        for(int i=0; i<n; ++i){
            rotation += A[i]*i;
            sum += A[i];
        }
        result = rotation;
        
        for(int i=0; i<n-1; ++i){
            rotation -= sum;
            rotation += long(A[i]) * n;
            result = max(result, rotation);
        }
        
        return result;
    }
};

File: Kotlin/396_Rotate_Function.kt
class Solution {
    fun maxRotateFunction(nums: IntArray): Int {
        if(nums.isEmpty()) return 0
        
        val n = nums.size
        var result = 0
        var sum = 0
        var rotation = 0
        
        for(i in 0 until n){
            rotation += nums[i] * i
            sum += nums[i]
        }
        result = rotation
        
        for(i in 0 until n - 1){
            rotation -= sum
            rotation += nums[i] * n
            result = maxOf(result, rotation)
        }
        
        return result
    }
}